language/move-prover/bytecode/src/borrow_analysis_v2.rs

```diff
@@ -0,0 +1,679 @@
+// Copyright (c) The Diem Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+//! Data flow analysis computing borrow information for preparation of memory_instrumentation.
+
+use crate::{
+    dataflow_analysis::{AbstractDomain, DataflowAnalysis, JoinResult, TransferFunctions},
+    function_target::{FunctionData, FunctionTarget},
+    function_target_pipeline::{FunctionTargetProcessor, FunctionTargetsHolder},
+    livevar_analysis::LiveVarAnnotation,
+    stackless_bytecode::{AssignKind, BorrowNode, Bytecode, Operation},
+    stackless_control_flow_graph::StacklessControlFlowGraph,
+};
+use itertools::Itertools;
+use move_model::{
+    ast::TempIndex,
+    model::{FunctionEnv, QualifiedId},
+};
+use std::collections::{BTreeMap, BTreeSet};
+use vm::file_format::CodeOffset;
+
+/// This enum denotes strong edges in the borrow graph
+#[derive(Debug, Clone, Eq, Ord, PartialEq, PartialOrd)]
+pub enum StrongEdge  {
+    Empty,
+    Offset(usize)
+}
+
+impl std::fmt::Display for StrongEdge {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            StrongEdge::Empty => write!(f, "E"),
+            StrongEdge::Offset(offset) => write!(f, "{}", offset),
+        }
+    }
+}
+
+pub enum BorrowEdge {
+    Weak,
+    Strong(StrongEdge)
+}
+
+/// This enum denotes the borrow graph edge domain.
+#[derive(Debug, Clone, Eq, Ord, PartialEq, PartialOrd)]
+pub enum EdgeDomain {
+    Weak,
+    Strong(BTreeSet<StrongEdge>)
```

Comment:
If you use `SetDomain<StrongEdge>` here instead, you should be able to simplify your join at line 68 by delegating to the join for `SetDomain`.

Reply:
Saw this just after latest push. I'll make these changes now.